### PR TITLE
revise neuronx-nemo-megatron example

### PR DESCRIPTION
**Pull request by KeitaW**
*Issue #, if available:*

*Description of changes:*

This PR is aiming to address some of the remaining comments @perifaws kindly made on the previous PR (https://github.com/aws-samples/awsome-distributed-training/pull/48).  

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by perifaws — CHANGES_REQUESTED** [review]

File: 3.test_cases/8.neuronx-nemo-megatron/2.setup-neuronx-nemo-megatron.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+APPS_PATH="$1"
+
+cd ${APPS_PATH}
+source ${APPS_PATH}/aws_neuron_venv_pytorch/bin/activate 
+git clone https://github.com/aws-neuron/neuronx-nemo-megatron.git
+cd neuronx-nemo-megatron
+pip3 install wheel
+./build.sh
+pip3 install ./build/*.whl
+pip3 install -r requirements.txt torch==1.13.1 protobuf==3.20.3
```
[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/8.neuronx-nemo-megatron/2.setup-neuronx-nemo-megatron.sh:12]: Actually, the `neuronx-nemo-megatron` is picky for the pytorch version. I would like to hard code the value until it relax the requirement.

File: 3.test_cases/8.neuronx-nemo-megatron/4.tokenize.sbatch [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+#SBATCH --exclusive
+#SBATCH --output=slurm-%x-%j.out
+#SBATCH --cpus-per-task 96
```
[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/8.neuronx-nemo-megatron/4.tokenize.sbatch:4]: The value was arbitrary. I fixed them to 128.

File: 3.test_cases/8.neuronx-nemo-megatron/4.tokenize.sbatch [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+python ${APPS_PATH}/neuronx-nemo-megatron/nemo/scripts/nlp_language_modeling/preprocess_data_for_megatron.py \
+    --input=${DATA_PATH}/book.jsonl \
+    --json-keys=text \
+    --tokenizer-library=huggingface \
+    --tokenizer-type=${MODEL_PATH}/Llama2-7b-hf \
+    --dataset-impl=mmap \
+    --output-prefix=${DATA_PATH}/book-tokenized \
+    --append-eod \
+    --need-pad-id \
+    --workers=32
```
[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/8.neuronx-nemo-megatron/4.tokenize.sbatch:20]: The value was arbitrary. I fixed them to 128.

File: 3.test_cases/8.neuronx-nemo-megatron/README.md [hunk 4]
```diff
@@ -54,16 +90,16 @@ Note: For the purposes of this sample we assume you have saved the Llama-2-7b mo
...
 This tutorial makes use of a [Red pyjama dataset](https://huggingface.co/datasets/togethercomputer/RedPajama-Data-1T). The dataset can be downloaded to your cluster by running the following commands on the head node:
 
-```
-mkdir -p /fsx/data/llama2
-wget https://data.together.xyz/redpajama-data-1T/v1.0.0/book/book.jsonl # Note: Dataset download is 50G and will take approximately 3-4 hours to download.
-or
-wget https://huggingface.co/datasets/togethercomputer/RedPajama-Data-1T-Sample/resolve/main/book_sample.jsonl -O /fsx/data/llama2/book.jsonl 
+```bash
+mkdir -p ${DATA_PATH} 
+wget https://data.together.xyz/redpajama-data-1T/v1.0.0/book/book.jsonl -O ${DATA_PATH}/book.jsonl # Note: Dataset download is 50G and will take approximately 3-4 hours to download.
```
[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/8.neuronx-nemo-megatron/README.md:95]: For quick testing,  we can always use `book_sample.jsonl` instead. I have added guidance for the sample data and faster download with `Aria2` in readme.

File: 3.test_cases/8.neuronx-nemo-megatron/5.precompile-model.sh [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+export APPS_PATH=/fsx
+export FSX_PATH=/fsx
+export MODEL_PATH=/fsx
+export DATA_PATH=$FSX_PATH/data/books
+export TEST_CASE_PATH=${APPS_PATH}/awsome-distributed-training/3.test_cases/8.neuronx-nemo-megatron  # where you copy the test case or set to your test case path
+cd ${APPS_PATH}/neuronx-nemo-megatron/nemo/examples/nlp/language_modeling
```
[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/8.neuronx-nemo-megatron/5.precompile-model.sh:2]: True, omitted.

File: 3.test_cases/8.neuronx-nemo-megatron/6.pretrain-model.sh [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+export APPS_PATH=/fsx
+export FSX_PATH=/fsx
+export MODEL_PATH=/fsx
+export DATA_PATH=$FSX_PATH/data/books
+export TEST_CASE_PATH=${APPS_PATH}/awsome-distributed-training/3.test_cases/8.neuronx-nemo-megatron  # where you copy the test case or set to your test case path
```
[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/8.neuronx-nemo-megatron/6.pretrain-model.sh]: True, omitted.

File: 3.test_cases/8.neuronx-nemo-megatron/README.md [hunk 7]
```diff
@@ -5,13 +5,39 @@
...
+
+```bash
+export APPS_PATH=/fsx
+export FSX_PATH=/fsx
+export MODEL_PATH=/fsx
+export DATA_PATH=$FSX_PATH/data/books
+export TEST_CASE_PATH=${APPS_PATH}/awsome-distributed-training/3.test_cases/8.neuronx-nemo-megatron  # where you copy the test case or set to your test case path
+```
+
+First of all, you need to have a Python virtual environment for `torch-neuronx` under `APPS_PATH`.
```
[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/8.neuronx-nemo-megatron/README.md]: added numbers

File: 3.test_cases/8.neuronx-nemo-megatron/README.md [hunk 8]
```diff
@@ -33,13 +59,23 @@ We will assume that you had placed the model and tokenizer as follows on cluster
 To convert the model to the standard Hugging Face format, the following script in transformers can be called with the following (example) command:
 
 ```
-sbatch 1.convert-weight.sbatch
+sbatch 3.convert-weight.sbatch
+```
+
+You can check progress of with `tail` command.
+
+```
```
[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/8.neuronx-nemo-megatron/README.md:66]: Thanks for catching this! updated

File: 3.test_cases/8.neuronx-nemo-megatron/README.md [hunk 9]
```diff
@@ -33,13 +59,23 @@ We will assume that you had placed the model and tokenizer as follows on cluster
...
+$ tail -f slurm-3.convert-weight.sbatch-xxx.out 
+Fetching all parameters from the checkpoint at /fsx/Llama2-meta/7B.
+Loading the checkpoint in a Llama model.
+Loading checkpoint shards: 100%|██████████| 33/33 [00:12<00:00,  2.65it/s]
+...
 ```
 
-Note: For the purposes of this sample we assume you have saved the Llama-2-7b model in a directory called `Llama2-7b-hf` with the following format:
+Once job completed, you will have the Llama-2-7b in a huggingface format under a directory called `Llama2-7b-hf` with the following format:
```
[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/8.neuronx-nemo-megatron/README.md]: Updated to 

> Once the job completed, you will have the Llama-2-7b model weights and tokenizer in a huggingface format under a directory called `Llama2-7b-hf` with the following format:

**Review by perifaws — APPROVED** [review]

**Comment by perifaws** [comment]
Merged, congrats and thanks @KeitaW !

**Pull request merged**